proposals/image-metadata.md

```diff
@@ -0,0 +1,81 @@
+
+We can add to these lists as we add more properties to the dev container configuration and the feature metadata.
+
+## Proposal
+
+We propose to add metadata to the image with the following structure using one entry per feature and devcontainer.json (see table below for the full list):
+
+```
+[
+	{
+		"id"?: string,
+		"init"?: boolean,
+		"privileged"?: boolean,
+		"capAdd"?: string[],
+		"securityOpt"?: string[],
+		"entrypoint"?: string,
+		"mounts"?: [],
+		...
+		"customizations"?: {
+			...
+		}
+	},
+	...
+]
+```
+
+To simplify adding this metadata for other tools, we also support having a single top-level object with the same properties.
+
+The metadata is added to the image as a `devcontainer.metadata` label with a JSON string value representing the above array or single object.
+
+## Merge Logic
+
+To apply the metadata together with a user's devcontainer.json at runtime the following merge logic by property is used. The table also notes which properties are currently supported coming from the devcontainer.json and which from the feature metadata, this will change over time as we add more properties.
+
+| Property | Type/Format | Merge Logic | devcontainer.json | Feature Metadata |
+| -------- | ----------- | ----------- | :---------------: | :--------------: |
+| `id` | E.g., `ghcr.io/devcontainers/features/node:1` | Not merged. |   | x |
+| `init` | `boolean` | `true` if at least one is `true`, `false` otherwise. |   | x |
+| `privileged` | `boolean` | `true` if at least one is `true`, `false` otherwise. |   | x |
+| `capAdd` | `string[]` | Union of all `capAdd` arrays without duplicates. |   | x |
+| `securityOpt` | `string[]` | Union of all `securityOpt` arrays without duplicates. |   | x |
+| `entrypoint` | `string` | Collected list of all entrypoints. |   | x |
+| `mounts` | `(string \| { type, src, dst })[]` | Collected list of all mountpoints. Conflicts: Last source wins. | x | x |
+| `onCreateCommand` | `string \| string[]` | Collected list of all onCreateCommands. | x |   |
+| `updateContentCommand` | `string \| string[]` | Collected list of all updateContentCommands. | x |   |
+| `postCreateCommand` | `string \| string[]` | Collected list of all postCreateCommands. | x |   |
+| `postStartCommand` | `string \| string[]` | Collected list of all postStartCommands. | x |   |
+| `postAttachCommand` | `string \| string[]` | Collected list of all postAttachCommands. | x |   |
+| `customizations` | Object of tool-specific customizations. | Merging is left to the tools. | x | x |
+| `remoteUser` | `string` | Last value wins. | x |   |
+| `userEnvProbe` | `string` (enum) | Last value wins. | x |   |
+| `remoteEnv` | Object of strings. | Per variable, last value wins. | x |   |
+| `containerEnv` | Object of strings. | Per variable, last value wins. | x |   |
```

Comment:
Isn't `containerEnv` supported in `Feature Metadata`? We do have it in `devcontainer-feature.json` in almost all of the official Features.

Reply:
Provided same `containerEnv` variable is present in `devcontainer.json` and feature metadata, should the value in `devcontainer.json` win ?

Reply:
Currently `containerEnv` from the feature metadata is added to the Dockerfile while `containerEnv` from the devcontainer.json is applied when the container starts. To also support variables like `${localEnv:VAR_NAME}` in the feature metadata's `containerEnv` it would make sense to apply these when the containers starts.

@edgonmsft @joshspicer We at one point moved the `ENV` statements in the Dockerfile before running the feature install scripts. Would moving these `ENV` statements out cause issues that cannot be addressed in a different way?

Reply:
We agreed to keep it as-is. 👍